Users db

dinara-urazova opened the pull request:
resolves #10 
Done
- added env_var for db_connection use, singleton
- added structure (user, db_schema), created db and users table
- separate file for funcs related to db (add_user, get_all_users, user_exists), save all users once they interact with db (add_user), check if the user exists in a db table (user_exists), find all users (get_all_users) 
- a secret command only for the bot's owner for sending newsletters, special offers etc

Commit:
add env_var to connect to db, create a table

Commit:
add broadcast feature (special offers, etc) which can be sent to all users

Commit:
format

Commit:
add env_var in env.ex

Commit:
add comment

Commit:
black format

Commit:
same again (head detached git bug, my bad)

Commit:
bloody black code (again)

Commit:
ruff check

Commit:
revert bug (removed import time by accident)

olegsklyarov commented on bot/__main__.py:
```diff
@@ -11,16 +14,34 @@
     six_months,
 )
 
+OWNER_CHAT_ID = 1059125420
```
Лучше вынеси в .env переменную

dinara-urazova replied:
done

olegsklyarov commented on create_postgresql_schema.py:
```diff
@@ -0,0 +1,29 @@
+
+connection = pg8000.connect(
+    user=env_config.postgresql_username,
+    password=env_config.postgresql_password,
```
`get_secret_value()`

olegsklyarov committed:
создание БД + логи в консоли

olegsklyarov commented on create_postgresql_schema.py:
```diff
@@ -1,29 +1,42 @@
+        first_name VARCHAR(255) NOT NULL,
+        last_name VARCHAR(255) NULL,
+        username VARCHAR(255) NULL,
+        created_at TEXT,
```
Тип данных должен быть timestamp, см https://www.postgresql.org/docs/current/datatype-datetime.html

dinara-urazova replied:
done

olegsklyarov commented on create_postgresql_schema.py:
```diff
@@ -1,29 +1,42 @@
+        last_name VARCHAR(255) NULL,
+        username VARCHAR(255) NULL,
+        created_at TEXT,
+        updated_at TEXT
```
Тут тоже timestamp

dinara-urazova replied:
done

Commit:
Changes
- add changes as per feedback (timestamp, rm owner as before)
- change owner_chat_id type to cf with int chat_id
- format with black

Commit:
ruff check fix

olegsklyarov commented on user_storage_postgresql.py:
```diff
@@ -0,0 +1,50 @@
+        just checks if a row exists, doesn't return any table data, can be unsafe (sql_injections) и дб параметризация (use %s as a placeholder, don't pass parameters directly), couldn't implement in pg8000.native
+        """
+        result = PostgreSQLSingleton.getConnection().run(
+            f"SELECT 1 FROM users WHERE telegram_id = {telegram_id}"
```
SELECT telegram_id ...

dinara-urazova replied:
done

olegsklyarov commented on user_storage_postgresql.py:
```diff
@@ -0,0 +1,50 @@
+        result = PostgreSQLSingleton.getConnection().run(
+            f"SELECT 1 FROM users WHERE telegram_id = {telegram_id}"
+        )
+        return bool(
```
return len(result) == 1

dinara-urazova replied:
немного по-другому сделала, так как не получилось через run передать параметры, поэтому используется cursor и проверка return cursor.fetchone() is not None

Commit:
rm pg8000.native, pass params in sql queries (for 'null' type to be shown instead of None in db)

olegsklyarov commented on user_storage_postgresql.py:
```diff
@@ -1,50 +1,64 @@
+        host=env_config.postgresql_hostname,
+        port=env_config.postgresql_port,
+        database=env_config.postgresql_database,
+    )
```
При такое реализации я не уверен, что произойдет закрытие соединения с БД. Даже если далее писать `with get_connection()...`

Вот так будет правильно:

```suggestion
    with pg8000.connect(
        user=env_config.postgresql_username,
        password=env_config.postgresql_password.get_secret_value(),
        host=env_config.postgresql_hostname,
        port=env_config.postgresql_port,
        database=env_config.postgresql_database,
    ) as connection:
        yield connection
```

olegsklyarov commented on user_storage_postgresql.py:
```diff
@@ -1,50 +1,64 @@
-            users.append(user)
-        print(users)
-        return users
+        with get_connection() as connection:
```
После правки в get_connection (см. комментарий выше), тут достаточно будет сделать:
```suggestion
        connection = get_connection()
```

olegsklyarov commented on user_storage_postgresql.py:
```diff
@@ -1,50 +1,64 @@
-        return bool(
-            result
-        )  # True if there's a user  - list with 1, False if None (empty list)
+        with get_connection() as connection:
```
Тут аналогично: не нужен `with`

dinara-urazova committed:
Update user_storage_postgresql.py

Co-authored-by: Oleg Sklyarov <oleg.sklyarov@gmail.com>

dinara-urazova committed:
Update user_storage_postgresql.py

Co-authored-by: Oleg Sklyarov <oleg.sklyarov@gmail.com>

Commit:
re feedback

olegsklyarov committed:
get_connection для использования в with ... as

olegsklyarov committed:
fix codestyle

olegsklyarov commented on user_storage_postgresql.py:
```diff
@@ -0,0 +1,69 @@
+                    updated_at=row[5],
+                )
+                users.append(user)
+                return users
```
Этот return выходит сразу после первой итерации. Нужно исправить отступ ❗️

dinara-urazova replied:
Поправила

Commit:
fix bug indentation
